Cast Integration Cleanup

OttoWinter opened the pull request:
## Description:

Clean up the cast integration. 

Most notably the `pychromecast.Chromecast` objects are now managed *by the entities*, not globally. This makes handling "moved" hosts, unavailable chromecasts and disconnecting *much* cleaner.

- [x] setup_platform and internal discovery now only create "ChromecastInfo" objects (with attrs 👌), not Chromecast objects. Missing information is automatically added via HTTP dial.
- [x] The CastDevice entities now take care of moved hosts themselves (using dispatcher callbacks).
- [x] Added `available` property.
- [x] Enabled polling mode **if an app is active**, because we sometimes don't receive media status callbacks from certain apps.
- [x] Added a ton of comments to hopefully make understanding the source code a bit easier.

Todo:
- [x] Some more local testing.
- [x] Fixing tests.
- [x] Add tests

Out of scope:
- [ ] Move some of the logic into pychromecast (would be good to test the logic with this user base first; but pychromecast could benefit from some logic, like availability, info objects, or half-polling mode; the last one is a bit controversial though)

fixes #13359, fixes #13319, fixes #13090

## Example entry for `configuration.yaml` (if applicable):
```yaml
media_player:
  - platform: cast
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

OttoWinter committed:
Cast Integration Cleanup

OttoWinter committed:
Fix long line

OttoWinter committed:
Fixes and logging

OttoWinter committed:
Fix tests

OttoWinter commented:
Tested extensively with chromecast, google home, cast audio and audio groups now. On to the tests!

OttoWinter committed:
Lint

MartinHjelmare commented on homeassistant/components/media_player/cast.py:
```diff
@@ -139,191 +190,441 @@ def _async_create_cast_device(hass, chromecast):
+        """Turn on the cast device."""
+        import pychromecast
+
+        if self._chromecast is None:
```
When can this happen? We set the cast info in `async_added_to_hass`.

OttoWinter replied:
When the chromecast "switched host". This happens for example when a cast audio group changes its "elected leader"/master. Then we first disconnect the old chromecast object and then connect the new one.

In that process, `self._chromecast` is briefly changed to `None` to indicate that the old chromecast object is invalid (any commands on the old object would result in exceptions, as the socket is no longer connected)

Additionally, when starting up, there might be a phase when the chromecast object is not connected yet but the entity is already visible in the frontend. This can be reproduced by manually specifying the cast device by `host:` in `configuration.yaml` and disabling any internet connection of your computer.

MartinHjelmare replied:
Shouldn't we set the entity to unavailable if it looses the connection to the chromecast? Then you wouldn't need to guard in all the methods.

It sounds weird that we allow entities to be created without a working connection during setup.

OttoWinter replied:
> Shouldn't we set the entity to unavailable if it looses the connection to the chromecast? 

Yes, that's also what this PR does. When `self._chromecast` is `None` available will also be `False`.

> Then you wouldn't need to guard in all the methods.

Ok, that's good. I put those checks everywhere to protect for potential race conditions, where the service calls are executed even though we're not connected. If the core can guarantee that these methods will never be called with an unavailable state I can also remove those checks.

> It sounds weird that we allow entities to be created without a working connection during setup.

That's true, and also was the previous behavior. The problem with that was that we would block the entire Home Assistant startup waiting for the socket client to connect. If no connection was available, we would even stop startup indefinitely - not good. I tried looking for a solution to that with `asyncio` for some time, but ultimately I think unavailable is better (and quicker).

I mean if we want to we can also try checking the HTTP API of the chromecast before adding the device, that would fix part of the problem (for standard chromecasts).

MartinHjelmare commented on homeassistant/components/media_player/cast.py:
```diff
@@ -139,191 +190,441 @@ def _async_create_cast_device(hass, chromecast):
         """Return the state of the player."""
-        if self.media_status is None:
+        if self._media_status is None:
             return STATE_UNKNOWN
```
Return `None` for unknown state.

MartinHjelmare commented on homeassistant/components/media_player/cast.py:
```diff
@@ -139,191 +190,441 @@ def _async_create_cast_device(hass, chromecast):
-        elif self.cast.is_idle:
+        elif self._chromecast is not None and self._chromecast.is_idle:
             return STATE_OFF
         return STATE_UNKNOWN
```
See above.

OttoWinter committed:
Report unknown state with None

OttoWinter committed:
Lint

OttoWinter committed:
Switch to async_add_job

Gets rid of those pesky "Setup of platform cast is taking over 10 seconds." messages.

edif30 commented:
Once ready let me know and I can test if you'd like.

OttoWinter committed:
Re-introduce PlatformNotReady

OttoWinter committed:
Add tests

OttoWinter committed:
Remove unnecessary checks

MartinHjelmare commented:
FYI, regarding core guarding from calling services for unavailable entities: Components that uses `async_extract_from_service` handles this. If a platform registers its own services it has to handle that on its own, in the service handler.

OttoWinter committed:
Test PlatformNotReady

OttoWinter committed:
Fix async in sync context

OttoWinter committed:
Blocking update

It's not using async anyway

OttoWinter reviewed (commented):
Added some comments here to point out some aspects of the changes that didn't really fit in code comments.

Since yesterday I also went through all sorts of scenarios of chromecasts disconnecting at certain times, no network connection, manually moving hosts around and so on. All of those "physical tests" appear to be working correctly with my set up + the `available` property now gives immediate feedback as to what is happening.

@edif30 If you want you could test these changes. It should essentially be as simple as copying [this file](https://raw.githubusercontent.com/OttoWinter/home-assistant/65deab10f11bd2406a87763f05ce8749cb92fdf0/homeassistant/components/media_player/cast.py) into `[config_dir]/custom_components/media_player/cast.py`.

MartinHjelmare commented on homeassistant/components/media_player/cast.py:
```diff
@@ -39,23 +41,103 @@
     vol.Optional(CONF_HOST): cv.string,
-    vol.Optional(CONF_IGNORE_CEC): [cv.string],
+    vol.Optional(CONF_IGNORE_CEC, default=[]): vol.All(cv.ensure_list,
+                                                       [cv.string])
```
LGTM.

OttoWinter committed:
Upgrade pychromecast to 2.1.0

OttoWinter committed:
Make reviewing easier

I like "protected" access, but I like reviewing more :)

OttoWinter committed:
Make reviewing even easier :)

OttoWinter committed:
Comment tests

balloob commented:
Awesome work 🐬

MaluNoPeleke commented:
@OttoWinter Is it already merged?
I just got a 

> Platform cast not ready yet. Retrying in 30 seconds.

See #13359

OttoWinter commented:
Yes it is in the beta. And that error message is expected now, it signals that one of Chromecast devices defined in your configuration.yaml can't be reached and therefore can't be setup.

MaluNoPeleke commented:
You mean the automatically created entity_registry.yaml?
Yesterday the same setup worked as expected but today not. All devices for the group are connected to power and the network (while not all devices of the list in the above mentioned file are connected to power).
Could the TTS and delay command be a problem?

```
    action:
      - service: media_player.volume_set
        data_template:
          entity_id: media_player.schlafzimmer
          volume_level: 0.1
      - service: tts.google_say
        data_template:
          entity_id: media_player.schlafzimmer
          message: Guten Morgen. Es ist {{now().hour}} Uhr {{now().minute}}
          cache: false
      - delay: 00:00:15
      - service: media_player.play_media
        data_template:
          entity_id: media_player.wecker
          media_content_id: 'http://fluxfm.hoerradar.de/fluxfmberlin-live-aac-mq?sABC=59s45n23%230%23no8617362n29q2o435p17n54928n16s5%23gharva&amsparams=playerid:tunein;skey:1509186083'
          media_content_type: 'audio/mp3'
```

OttoWinter commented:
1. What are you using to discover your cast devices? `discovery:` in your configuration.yaml, or `media_player: platform: cast` or manually by `host:`?
2. In what way are they not working? Is it just the error message or do they not turn on/execute the command?
3. If you have some Chromecast devices in your **configuration.yaml** that are **not** connected to power that error *message* is expected; the other devices should work correctly though. (I hope, at least they did during my testing)
4. The TTS and delay shouldn't really affect anything here. The entity_regitry.yaml neither.
5. I don't think this discussion should be done in this PR, as it's not a comment on the code. Please open a separate issue and I will help you out there.
